### PR TITLE
Move compiler version macros into libcu++

### PR DESCRIPTION
**Pull request by bernhardmgruber**
This PR deprecates CUB and Thrust compiler detection and version macros. The version macros are also added into libcu++.

Fixes: #1965

### CODE REVIEW

File: thrust/thrust/random/detail/normal_distribution_base.h [hunk 1]
```diff
@@ -148,7 +148,7 @@ class normal_distribution_portable
 template <typename RealType>
 struct normal_distribution_base
 {
-#if THRUST_DEVICE_COMPILER == THRUST_DEVICE_COMPILER_NVCC && !defined(_NVHPC_CUDA)
+#if defined(_CCCL_CUDA_COMPILER) && !defined(_NVHPC_CUDA)
```
Comment (miscco) [inline thrust/thrust/random/detail/normal_distribution_base.h:151]:
```suggestion
#if defined(_CCCL_CUDA_COMPILER) && !defined(_CCCL_CUDA_COMPILER_NVHPC)
```

File: libcudacxx/include/cuda/std/__cccl/compiler.h [hunk 2]
```diff
@@ -20,10 +20,14 @@
...
 #  define _CCCL_COMPILER_CLANG
+#  define _CCCL_CLANG_VERSION (__clang_major__ * 10000 + __clang_minor__ * 100 + __clang_patchlevel__)
 #elif defined(__GNUC__)
 #  define _CCCL_COMPILER_GCC
+#  define _CCCL_GCC_VERSION (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__)
 #elif defined(_MSC_VER)
 #  define _CCCL_COMPILER_MSVC
+#  define _CCCL_MSVC_VERSION      _MSC_VER
+#  define _CCCL_MSVC_VERSION_FULL _MSC_FULL_VER
 #elif defined(__IBMCPP__)
```
Comment (fbusato) [inline libcudacxx/include/cuda/std/__cccl/compiler.h:31]: We don't support PowerPC/ppc64le/IBM platform anymore

Comment (fbusato) [reply libcudacxx/include/cuda/std/__cccl/compiler.h:31]: Also, Intel ICC is deprecated

Comment (bernhardmgruber) [reply libcudacxx/include/cuda/std/__cccl/compiler.h:31]: Then we should probably emity a warning when a user attempts tp compile for such a platform. Can you please create an issue to track that? Also, please include a reference where we state that support is no longer provided. Thx!

Comment (miscco) [reply libcudacxx/include/cuda/std/__cccl/compiler.h:31]: I believe that is just coming from libc++ not the toolkit

File: thrust/thrust/complex.h [hunk 3]
```diff
@@ -76,7 +76,7 @@ namespace detail
 template <typename T, std::size_t Align>
 struct complex_storage;
 
-#if THRUST_CPP_DIALECT >= 2011 && (THRUST_HOST_COMPILER == THRUST_HOST_COMPILER_GCC) && (THRUST_GCC_VERSION >= 40800)
+#if THRUST_CPP_DIALECT >= 2011 && defined(_CCCL_COMPILER_GCC) && (_CCCL_GCC_VERSION >= 40800)
```
Comment (miscco) [inline thrust/thrust/complex.h:79]: Nitpick: This should be either
```suggestion
#if defined(_CCCL_COMPILER_GCC) && (_CCCL_GCC_VERSION >= 40800)
```

or
```suggestion
#if _CCCL_STD_VER >= 2011 && defined(_CCCL_COMPILER_GCC) && (_CCCL_GCC_VERSION >= 40800)
```

File: thrust/thrust/detail/complex/catrig.h [hunk 4]
```diff
@@ -671,7 +671,7 @@ _CCCL_HOST_DEVICE inline double real_part_reciprocal(double x, double y)
  * Re(catanh(z)) = x/|z|^2 + O(x/z^4)
  *    as z -> infinity, uniformly in x
  */
-#if THRUST_CPP_DIALECT >= 2011 || THRUST_HOST_COMPILER != THRUST_HOST_COMPILER_MSVC
+#if _CCCL_STD_VER >= 2011 || !defined(_CCCL_COMPILER_MSVC)
```
Comment (miscco) [inline thrust/thrust/detail/complex/catrig.h:674]: Can we just drop all those? We require C++11 so there is no harm is just dropping the noise

File: thrust/thrust/random/detail/normal_distribution.inl [hunk 5]
```diff
@@ -33,7 +33,7 @@
 #include <cstdint>
 
 // for floating point infinity
-#if THRUST_DEVICE_COMPILER == THRUST_DEVICE_COMPILER_NVCC
+#ifdef _CCCL_CUDA_COMPILER
```
Comment (miscco) [inline thrust/thrust/random/detail/normal_distribution.inl:36]: Nitpick: This should just be `#include <cuda/std/limits>`. Feel free to open an issue

**Review by miscco — APPROVED** [review]
Some minor nitpicks. Feel free to convert into issues for the additional cleanup

File: thrust/thrust/complex.h (hunk 3 above)
Comment (bernhardmgruber) [reply thrust/thrust/complex.h:79]: I went for a slighly bigger refactoring and dropped everything for the plain C++11 implementation.

File: thrust/thrust/detail/complex/catrig.h (hunk 4 above)
Comment (bernhardmgruber) [reply thrust/thrust/detail/complex/catrig.h]: Dropped.

**Pull request merged**